tutorbackup/templates/backup/build/backup/backup_services.py

```diff
@@ -0,0 +1,186 @@
+    outfile = MYSQL_DUMPFILE
+
+    logger.info(f"Dumping MySQL databases on {host}:{port} to {outfile}")
+    cmd = ("mysqldump "
+           "--all-databases --add-drop-database --routines "
+           "--events --single-transaction "
+           "--quick --quote-names --max-allowed-packet=16M "
+           f"--host={host} --port={port} "
+           f"--user={user} --password={password}")
+    with open(outfile, 'wb') as out:
+        check_call(cmd,
+                   shell=True,
+                   stdout=out,
+                   stderr=sys.stderr)
+
+    size = os.path.getsize(outfile)
+    logger.info(f"Complete. {outfile} is {size} bytes.")
+
+
+def mongodump():
+    host = ENV['MONGODB_HOST']
+    port = ENV['MONGODB_PORT']
+    outdir = MONGODB_DUMPDIR
+
+    logger.info(f"Dumping MongoDB databases on {host}:{port} to {outdir}")
+    cmd = ("mongodump "
+           f"--out={outdir} "
+           f"--host={host} "
+           f"--port={port}")
+    try:
+        cmd += (f" --username={ENV['MONGODB_USERNAME']} "
+                f"--password={ENV['MONGODB_PASSWORD']}")
+    except KeyError:
+        pass
+
+    check_call(cmd,
+               shell=True,
+               stdout=sys.stdout,
+               stderr=sys.stderr)
+
+    total_size = get_size(outdir)
+    logger.info(f"Complete. {outdir} total size {total_size} bytes.")
+
+
+def caddydump():
+    outdir = CADDY_DUMPDIR
+    logger.info(f"Copying Caddy data to {outdir}")
+    shutil.copytree('caddy', outdir)
+
+    total_size = get_size(outdir)
+    logger.info(f"Complete. {outdir} total size {total_size} bytes.")
+
+
+def archive():
+    paths = [MYSQL_DUMPFILE, MONGODB_DUMPDIR]
+    if PERFORM_CADDY_BACKUP:
+        paths.append(CADDY_DUMPDIR)
+
+    outfile = TARFILE
+
+    logger.info(f"Creating archive {outfile}")
+    with tarfile.open(outfile, "w:xz") as tar:
+        for item in paths:
+            tar.add(item)
+
+    size = os.path.getsize(outfile)
+    logger.info(f"Complete. {outfile} is {size} bytes.")
+
+
+def upload_to_s3():
+    import hashlib
+    from s3_client import S3_CLIENT
+
+    bucket = ENV['S3_BUCKET_NAME']
+    file_name = TARFILE
+
+    logger.info(f"Uploading {file_name} to S3 bucket {bucket}")
+    try:
+        S3_CLIENT.upload_file(
+            file_name,
+            bucket,
+            os.path.basename(file_name),
+        )
+        logger.info(f"Uploaded {file_name} to {bucket}.")
+
+        logger.info("Checking uploaded file's integrity ...")
+        obj_metadata = S3_CLIENT.head_object(
+            Bucket=bucket,
+            Key=os.path.basename(file_name),
+        )
+
+        received_checksum = obj_metadata['ETag'].strip('"')
+        calculated_checksum = hashlib.md5(
+            open(file_name, 'rb').read()).hexdigest()
+        version_id = obj_metadata['VersionId']
+        size = obj_metadata['ContentLength']
+
+        if received_checksum == calculated_checksum:
+            logger.info("File integrity verified.")
```

Comment:
What we're logging here is a single event. So I think it's better to make this one message, rather than 4.

Reply:
I'll change that